R/pd_cpfw_merge.R

```diff
@@ -0,0 +1,497 @@
+#' Merge country/survey PFW info with dataliweb survey data
+#'
+#' @param dt DLW country/survey data
+#' @param pfw PFW
+#'
+#' @return list
+#' @export
+#'
+#' @examples
+#' pfw  <- pipload::pip_load_aux("pfw")
+#' md   <- pipload::pip_load_dlw(country = "PHL", 2012)
+#' l    <- pd_cpfw_merge(md, pfw)
+#'
+#' gd   <- pipload::pip_load_dlw("CHN", 2015)
+#' l    <- pd_cpfw_merge(gd, pfw)
+pd_cpfw_merge <- function(dt, pfw) {
+
+  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+  # computations   ---------
+  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+  ## PFW for specific country --------
+
+  cpfw <- get_country_pfw(dt, pfw)
+
+  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+  ## Split alternative welfare --------
+
+  lf   <- pd_split_alt_welfare(dt, cpfw)
+
+  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+  ## Map survey to cpfw --------
+
+  lfs <- purrr::map2(.x = lf,
+                     .y =  cpfw,
+                     .f = cpfw_merge)
+
+  # names(lfs) <- sapply(cpfw, `[[`, "cache_id") #Maybe not needed
+  # lfs <- Filter(function(df) !all(is.na(df)), lfs)
+
+  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+  # Return   ---------
+  #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+  return(lfs)
+
+}
+
+#' Merge country/survey PFW with dataliweb survey data
+#'
+#' @param dt data.table
+#' @param cpfw data.table with country/survey PFW
+#' @param ...  other parameters
+#'
+#' @return data.table
+#' @export
+cpfw_merge <- function(dt, cpfw, ...){
+
+      # Create hard copy
+      dt_c <- copy(dt)
+
+      #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+      ## Main variables (same for md and gd) --------
+      dt_c <- add_main_vars(dt_c, cpfw)
+
+      #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+      ## Area (Needed for Domain variables)--------
+      dt_c <- add_area(dt_c)
+
+      #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+      ## Domain variables (same for md and gd) --------
+      dt_c <- add_dom_vars(dt_c, cpfw)
+
+      #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+      ## Distribution type  (different for md and gd) --------
+      dt_c <- add_dist_type(dt_c, cpfw)
+
+      #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+      ## Transform unique variables into attributes --------
+
+      dt_f <- col_to_attr(dt_c, cpfw)
+
+
+  return(dt_f)
+
+}
+
+#' Add metadata variables to country/survey
+#'
+#' @inheritParams cpfw_merge
+#'
+#' @return data.table
+#'
+#' @keywords internal
+add_main_vars <- function(dt, cpfw) {
+
+  tryCatch(
+    expr = {
+
+      #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+      # computations   ---------
+      #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+      variables <- colnames(dt)
+
+      main_vars <- c("survey_year",
+                     "country_code",
+                     "welfare_type")
+
```

Comment:
I would add this list as an option predifined in `zzz.R` file and, unless you have reason for the current order, please place country_code first, then survey_year, then reporting_level (which is not there and I don't know if you may need it here) and, finally, welfare_type.